### DISCUSSION TITLE
How to configure custom features with Rails 7.1 + Zeitwerk

### DISCUSSION THREAD

**Discussion by renchap** [janko/rodauth-rails#256] [Q&A] [answered]
I am upgrading my app to Rails 7.1, and I am having Zeitweik complaining about a custom feature not defining an expected class:

`Zeitwerk::NameError: expected file lib/rodauth_verify_account_forever_grace.rb to define constant RodauthVerifyAccountForeverGrace, but didn't`

The file is defining a standard feature:

```ruby
module Rodauth
  Feature.define(:verify_account_forever_grace, :VerifyAccountForeverGrace) do
    …
  end
end
```

I read https://github.com/janko/rodauth-rails/wiki/Custom-Extensions but I am not sure how this can work for non-trivial features? For example one of ours calls `depends` and `session_key` outside of class methods, but those are not defined is the feature is a plain module.

I have not been able to find documentation on how to create complex features without using `Feature.define`, is this even possible?

**Comment by janko** [comment, marked as answer]
I'm assuming you're getting this error with the new `config.autoload_lib` setting. The way to make Zeitwerk happy is move the feature to `lib/rodauth/verify_account_forever_grace.rb`. This satisfies Zeitwerk's file naming requirements, because Rodauth internally defines a `Rodauth::VerifyAccountForeverGrace` module.

Because it's not in `rodauth/features/*` directory that Rodauth will automatically require from, you just need to manually `require "rodauth/verify_account_forever_grace"` before enabling the feature. But I'm assuming you're already doing that 🙂 It seems that code reloading for this file still works even if you manually require it.

I will add these instructions to the wiki page.

**Reply by renchap** [reply]
Thanks, that is working well and solved my issue. I tried to understand the meta-programming instantiated the feature but I did not get that it created a class with this name.
